### PR TITLE
update-search-index

### PR DESCRIPTION
**Pull request by josh8551021**
This pull reuquest introduces the ElasticsearchClient which can currently be used to add and update restaurant headers in Elasticsearch.

First commit: 0438fbd2

### CODE REVIEW

File: capstone/pom.xml [hunk 1]
```diff
@@ -88,6 +88,12 @@
       <artifactId>mockito-core</artifactId>
       <version>3.4.0</version>
     </dependency>
+
+    <dependency>
+      <groupId>com.google.http-client</groupId>
+      <artifactId>google-http-client-gson</artifactId>
+      <version>1.36.0</version>
+    </dependency>
```
Comment (Bickor) [inline capstone/pom.xml:96]: Isn't that already in there? Just checking in case it might be duplicated. If it is not then ignore this comment.

Comment (josh8551021) [reply capstone/pom.xml:96]: I think there was a similar dependency for just the HTTP client. That dependency was needed for specifically the gson factory which it turns out I'm not using, haha. I'll remove that dependency.

File: capstone/src/main/java/com/google/step/search/ElasticsearchClient.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    try {
+      HttpRequest request = requestFactory.buildPutRequest(requestUrl, putRequestContent);
+      HttpResponse response = request.execute();
+
+      statusCode = response.getStatusCode();
+    } catch (IOException e) {
+      statusCode = HttpURLConnection.HTTP_INTERNAL_ERROR;
+    }
+
+    return statusCode;
```
Comment (tzsebe) [inline capstone/src/main/java/com/google/step/search/ElasticsearchClient.java:92]: Final error-related comment, and probably the one you should really be doing: Consider just throwing HttpResponseException, IOException, and letting the client handle it as they see fit. Just a status code isn't really enough information, so passing up the full info is best.

File: capstone/src/main/java/com/google/step/search/ElasticsearchClient.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+  private static final String elasticsearchHostname = "10.128.0.2";
+  private static final short elasticsearchPort = 9200;
+  private final String elasticsearchUriString;
+
+  private static final String RESTAURANTS = "restaurants";
+
+  private final HttpRequestFactory requestFactory;
+
+  Gson gson = new Gson();
```
[the remark this answers is not in the document]
Comment (josh8551021) [reply capstone/src/main/java/com/google/step/search/ElasticsearchClient.java:49]: Oh, I hadn't realized that it wasn't. I'll change that.

File: capstone/src/main/java/com/google/step/search/ElasticsearchClient.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+
+/**
+ * The {@code ElasticsearchClient} is a class that sends requests to add {@code RestaurantHeaders}
+ * to the "restaurants" index of Elasticsearch.
+ */
+public class ElasticsearchClient {
+
+  private static final String elasticsearchHostname = "10.128.0.2";
```
[the remark this answers is not in the document]
Comment (josh8551021) [reply capstone/src/main/java/com/google/step/search/ElasticsearchClient.java:41]: Will do.

File: capstone/src/main/java/com/google/step/data/RestaurantHeader.java [hunk 5]
```diff
@@ -47,9 +47,10 @@ private RestaurantHeader(long restaurantKey,
    * @param restaurant The {@code Restaurant} from which the header information would come
    * @return A newly created RestaurantHeader object
    */
-  public static RestaurantHeader createRestaurantHeader(Restaurant restaurant) {
+
+  public static RestaurantHeader createHeaderFromRestaurant(Restaurant restaurant) {
     return new RestaurantHeader(
-        restaurant.getRestaurantKey(),
+        restaurant.getRestaurantKey().orElse(0L),
```
[the remark this answers is not in the document]
Comment (josh8551021) [reply capstone/src/main/java/com/google/step/data/RestaurantHeader.java:53]: It's the case of the former. I'll add the assertion.

File: capstone/src/main/java/com/google/step/search/ElasticsearchClient.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    requestFactory = transport.createRequestFactory();
+
+    URIBuilder uriBuilder = new URIBuilder()
+        .setScheme("http")
+        .setHost(elasticsearchHostname)
+        .setPort(elasticsearchPort);
+    elasticsearchUriString = uriBuilder.toString();
+  }
+
+  public ElasticsearchClient() {
```
[the remark this answers is not in the document]
Comment (josh8551021) [reply capstone/src/main/java/com/google/step/search/ElasticsearchClient.java:61]: I'll make those changes.

File: capstone/src/main/java/com/google/step/search/ElasticsearchClient.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    HttpContent putRequestContent = new ByteArrayContent(Json.MEDIA_TYPE,
+        requestBody.getBytes(StandardCharsets.UTF_8));
+
+    int statusCode;
+    try {
+      HttpRequest request = requestFactory.buildPutRequest(requestUrl, putRequestContent);
+      HttpResponse response = request.execute();
+
+      statusCode = response.getStatusCode();
+    } catch (IOException e) {
```
[the remark this answers is not in the document]
Comment (josh8551021) [reply capstone/src/main/java/com/google/step/search/ElasticsearchClient.java:83]: Should I be doing this or just letting the exception propagate to the front end?

File: capstone/src/main/java/com/google/step/search/ElasticsearchClient.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (josh8551021) [reply capstone/src/main/java/com/google/step/search/ElasticsearchClient.java]: Done.

File: capstone/src/main/java/com/google/step/search/ElasticsearchClient.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (josh8551021) [reply capstone/src/main/java/com/google/step/search/ElasticsearchClient.java]: Done.

File: capstone/src/main/java/com/google/step/search/ElasticsearchClient.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (josh8551021) [reply capstone/src/main/java/com/google/step/search/ElasticsearchClient.java]: Done.

File: capstone/src/main/java/com/google/step/search/ElasticsearchClient.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (josh8551021) [reply capstone/src/main/java/com/google/step/search/ElasticsearchClient.java:83]: I decided to remove the error handling. Let me know if you think I should do something different.

File: capstone/src/main/java/com/google/step/search/ElasticsearchClient.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+    GenericUrl requestUrl = new GenericUrl(elasticsearchUriString);
+    requestUrl.setPathParts(Arrays.asList("", RESTAURANTS, "_doc", restaurantKey));
+
+    String requestBody = gson.toJson(restaurantHeader);
+    HttpContent putRequestContent = new ByteArrayContent(Json.MEDIA_TYPE,
+        requestBody.getBytes(StandardCharsets.UTF_8));
+
+    requestFactory.buildPutRequest(requestUrl, putRequestContent).execute();
+
```
Comment (mkmg) [inline capstone/src/main/java/com/google/step/search/ElasticsearchClient.java:74]: Nit: Remove blank line at end of method.

**Pull request merged**